experiment/simulation/js/gate.js

```diff
@@ -0,0 +1,317 @@
+    gate.registerComponent("working-area");
+}
+
+window.addGate = addGate;
+
+export function getResult(gate) {
+    if (gate.output != null) {
+        return;
+    }
+    for (let i = 0; i < gate.inputs.length; i++) {
+        if (gate.inputs[i].output == null) {
+            getResult(gate.inputs[i]);
+        }
+    }
+    gate.generateOutput();
+    return;
+}
+
+function setInput(event) {
+    let parentElement = event.target.parentElement;
+    let element = event.target;
+    let type = parentElement.className.split(" ")[0];
+    let gate = gates[parentElement.id];
+    if (type === "high") {
+        // change class HIGH to LOW
+        parentElement.classList.replace("high", "low");
+        element.innerHTML = "0";
+        gate.setOutput(false);
+    }
+    else if (type === "low") {
+        parentElement.classList.replace("low", "high");
+        element.innerHTML = "1";
+        gate.setOutput(true);
+    }
+}
+
+window.setInput = setInput;
+
+export function clearResult() {
+    const result = document.getElementById("result");
+    result.innerHTML = "";
+}
+
+export function printErrors(message,objectId) {
+    const result = document.getElementById('result');
+    result.innerHTML += message;
+    result.className = "failure-message";
+    if(objectId !== null)
+    {
+        objectId.classList.add("highlight")
+        setTimeout(function () {objectId.classList.remove("highlight")}, 5000);
+    }
+}
+
+// Check if the connections are correct
+export function checkConnections() {
+    for (let gateId in gates) {
+        const gate = gates[gateId];
+        const id = document.getElementById(gate.id);
+        if (gate.inputPoints.length != gate.inputs.length) {
+            printErrors("Highlighted component not connected properly\n",id);
+            return false;
+        } else if (gate.isConnected === false && gate.isOutput === false) {
+            printErrors("Highlighted component not connected properly\n",id);
+            return false;
+        }
+    }
+    return true;
+}
+
+// Simulate the circuit
+export function simulate() {
+    clearResult();
+    if (!checkConnections()) {
+        return;
+    }
+
+    // reset output in gate
+    for (let gateId in gates) {
+        if (!gates[gateId].isInput) {
+            gates[gateId].output = null;
+        }
+    }
+
+    for (let gateId in gates) {
+        const gate = gates[gateId];
+        if (gate.isOutput) {
+            getResult(gate);
+            let element = document.getElementById(gate.id);
+            if (gate.output) {
+                element.className = "high";
+                element.childNodes[0].innerHTML = "1";
+            } else {
+                element.className = "low";
+                element.childNodes[0].innerHTML = "0";
+            }
+        }
+    }
+}
+
+window.simulate = simulate;
+
+// Simulate the circuit for given gates; Used for testing the circuit for all possible inputss
+export function testSimulation(gates) {

```

Comment:
You do not need to change it here, but for future reference, the testing code usually goes in separate files.

Reply:
sure, will make note of that